### PR TITLE
[WIP]Support Uniffle columnar shuffle

### PR DESCRIPTION
**Pull request by summaryzb**
## What changes were proposed in this pull request?
Support Uniffle columnar shuffle

## How was this patch tested?
Integration Testing with Uniffle in Velox CI

### CODE REVIEW

File: gluten-uniffle/src/main/java/org/apache/spark/shuffle/gluten/uniffle/UniffleShuffleManager.java [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+  // and Spark driver will make some decision according to coordinator,
+  // e.g. determining what RSS servers to use.
+  // Then Spark driver will return a ShuffleHandle and
+  // pass that ShuffleHandle to executors (getWriter/getReader)
+  @Override
+  public <K, V, C> ShuffleHandle registerShuffle(
+      int shuffleId, ShuffleDependency<K, V, C> dependency) {
+    if (!(dependency instanceof ColumnarShuffleDependency)) {
+      columnarShuffleIds.add(shuffleId);
+      return columnarShuffleManager().registerShuffle(shuffleId, dependency);
```
Comment (loukey-lj) [inline gluten-uniffle/src/main/java/org/apache/spark/shuffle/gluten/uniffle/UniffleShuffleManager.java:106]: why not super.registerShuffle(shuffleId, dependency);

File: gluten-uniffle/src/main/java/org/apache/spark/shuffle/gluten/uniffle/UniffleShuffleManager.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    } else {
+      writeMetrics = context.taskMetrics().shuffleWriteMetrics();
+    }
+    WriteBufferManager bufferManager = new WriteBufferManager(
+        shuffleId, taskId, context.taskAttemptId(), bufferOptions,
+        null,
+        rssHandle.getPartitionToServers(), context.taskMemoryManager(),
+        writeMetrics, RssSparkConfig.toRssConf(sparkConf), this::sendData);
+    LOG.info("RssHandle appId {} shuffleId {} ", rssHandle.getAppId(), rssHandle.getShuffleId());
+    return new UniffleColumnarShuffleWriter<>(
```
Comment (loukey-lj) [inline gluten-uniffle/src/main/java/org/apache/spark/shuffle/gluten/uniffle/UniffleShuffleManager.java:142]: why not  UniffleColumnarShuffleWriter(super.getWriter(),...) .   Too much RSS code is exposed in gluten

File: gluten-uniffle/src/main/java/org/apache/spark/shuffle/gluten/uniffle/UniffleShuffleManager.java [hunk 3]
```diff
@@ -0,0 +1,874 @@
...
+import org.apache.uniffle.common.util.RssUtils;
+import org.apache.uniffle.common.util.ThreadUtils;
+import org.apache.uniffle.shuffle.manager.RssShuffleManagerBase;
+import org.apache.uniffle.shuffle.manager.ShuffleManagerGrpcService;
+import org.apache.uniffle.shuffle.manager.ShuffleManagerServerFactory;
+
+import static org.apache.uniffle.common.config.RssBaseConf.*;
+import static org.apache.uniffle.common.config.RssClientConf.MAX_CONCURRENCY_PER_PARTITION_TO_WRITE;
+
+public class UniffleShuffleManager extends RssShuffleManagerBase {
```
[the remark this answers is not in the document]
Comment (summaryzb) [reply gluten-uniffle/src/main/java/org/apache/spark/shuffle/gluten/uniffle/UniffleShuffleManager.java]: I'm not sure whether gluten with rss support the dynamic feature concurrently.
even though, maybe it's better to change uniffle side for compatible

File: gluten-uniffle/src/main/java/org/apache/spark/shuffle/gluten/uniffle/UniffleShuffleManager.java (hunk 1 above)
Comment (summaryzb) [reply gluten-uniffle/src/main/java/org/apache/spark/shuffle/gluten/uniffle/UniffleShuffleManager.java:107]: Maybe it's present in next line

File: gluten-uniffle/src/main/java/org/apache/spark/shuffle/gluten/uniffle/UniffleShuffleManager.java (hunk 2 above)
Comment (summaryzb) [reply gluten-uniffle/src/main/java/org/apache/spark/shuffle/gluten/uniffle/UniffleShuffleManager.java:141]: Sounds reasonable, but  `rss.row.based` have to be set in `WriteBufferManager`, Though it looks really not elegant

**Comment by weiting-chen** [comment]
There are some fails in the UT testing during compile phase, could you help to check and rebase the code?

**Comment by LuciferYang** [comment]
Why was this PR closed?

**Comment by rhh777** [comment]
+1

**Pull request closed without merging**